osbrain/tests/test_agent_sync_publications.py

```diff
@@ -73,6 +73,41 @@ def on_error(agent):
     agent.error_log.append('error')
 
 
+@pytest.mark.parametrize('socket_type', ['PUB', 'SYNC_PUB'])
+def test_change_subscription_topics_sync(nsproxy, socket_type):
+    '''
+    Test for the different options of subscribing/unsubscribing to topics
+    in the SYNC_PUB/SYNC_SUB pattern.
+    '''
+    server = run_agent('server')
+    client = run_agent('client')
+
+    addr = server.bind(socket_type, alias='pub', handler=lambda: None)
+    client.set_attr(received=[])
+    client.connect(addr, alias='sub', handler=append_received)
+
+    # Stablish a connection
+    server.each(0.1, 'send', 'pub', 'connecting...')
+    assert wait_agent_attr(client, name='received', data='connecting...')
+
+    # By default, client will be subscribed to all topics
+    server.send('pub', 'hello')
+    assert wait_agent_attr(client, name='received', data='hello')
+
+    # Only subscribe to 'TOP' topic
+    client.unsubscribe_socket_from_topic('sub', b'')
+    client.subscribe_socket_to_topic('sub', b'TOP')
+
+    # Message not received since 'TOP' topic not specified in the send call
+    server.send('pub', 'world')
```

Comment:
Again we find synchronization problems here.

What do you think about this? We create a function only for tests `sync_pub_sub_append_received` (for example). This function takes a publisher and a subscriber. Assumes the handler in the subscriber is the classic `append_received` (which we use most of the time). Starts publishing messages like increasing natural numbers (in a for loop, not using a timer, we can set a small sleep here). We wait for the first one to be appended in the client. Then we no longer send messages from the publisher. We wait for the client to receive the last number sent from the publisher (once the first is received, all messages after that one should also be received, and we know which number we sent last from the publisher). Then PUB-SUB are definitely connected and in sync. Now we reset `received` in the client and there we go!

What you think?

(This would apply for other tests we have around there, until we implement heart-beating or some other fancy stuff).